File: planner-api/src/infrastructure/redis/redis.service.ts
import { Injectable, OnModuleDestroy } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import Redis from 'ioredis';

@Injectable()
export class RedisService implements OnModuleDestroy {
    private readonly redis: Redis;

    constructor(private readonly configService: ConfigService) {
        this.redis = new Redis({
            host: this.configService.get('REDIS_HOST', 'localhost'),
            port: this.configService.get('REDIS_PORT', 6379),
            password: this.configService.get('REDIS_PASSWORD'),
            maxRetriesPerRequest: 3,
        });

        this.redis.on('error', (error) => {
            console.error('Redis connection error:', error);
        });

        this.redis.on('connect', () => {
            console.log('✅ Redis connected successfully');
        });
    }

    async onModuleDestroy() {
        await this.redis.quit();
    }

    async set(key: string, value: any, ttl?: number): Promise<void> {
        const serializedValue = JSON.stringify(value);
        if (ttl) {
            await this.redis.setex(key, ttl, serializedValue);
        } else {
            await this.redis.set(key, serializedValue);
        }
    }

    async get<T>(key: string): Promise<T | null> {
        const value = await this.redis.get(key);
        if (!value) return null;
        return JSON.parse(value);
    }

    async del(key: string): Promise<void> {
        await this.redis.del(key);
    }

    async exists(key: string): Promise<boolean> {
        const result = await this.redis.exists(key);
        return result === 1;
    }

    async expire(key: string, seconds: number): Promise<void> {
        await this.redis.expire(key, seconds);
    }

    async publish(channel: string, message: any): Promise<void> {
        await this.redis.publish(channel, JSON.stringify(message));
    }

    async subscribe(channel: string, callback: (message: any) => void): Promise<void> {
        const subscriber = this.redis.duplicate();
        await subscriber.subscribe(channel);

        subscriber.on('message', (ch, message) => {
            if (ch === channel) {
                callback(JSON.parse(message));
            }
        });
    }

    async healthCheck(): Promise<boolean> {
        try {
            await this.redis.ping();
            return true;
        } catch (error) {
            return false;
        }
    }
} 

File: ml-service/app/services/redis_service.py
import redis.asyncio as redis
import os
from typing import Optional

class RedisService:
    def __init__(self):
        self.redis_client: Optional[redis.Redis] = None
        self.host = os.getenv("REDIS_HOST", "localhost")
        self.port = int(os.getenv("REDIS_PORT", 6379))

    async def connect(self):
        """Connect to Redis"""
        try:
            self.redis_client = redis.Redis(
                host=self.host,
                port=self.port,
                decode_responses=True
            )
            await self.redis_client.ping()
            print(f"✅ Connected to Redis at {self.host}:{self.port}")
        except Exception as e:
            print(f"❌ Redis connection failed: {e}")
            self.redis_client = None

    async def disconnect(self):
        """Disconnect from Redis"""
        if self.redis_client:
            await self.redis_client.close()
            print("🔌 Disconnected from Redis")

    async def ping(self) -> bool:
        """Ping Redis server"""
        try:
            if self.redis_client:
                await self.redis_client.ping()
                return True
            return False
        except:
            return False

    async def set(self, key: str, value: str, expire: int = 3600):
        """Set key-value pair with expiration"""
        if self.redis_client:
            await self.redis_client.set(key, value, ex=expire)

    async def get(self, key: str) -> Optional[str]:
        """Get value by key"""
        if self.redis_client:
            return await self.redis_client.get(key)
        return None

    async def delete(self, key: str):
        """Delete key"""
        if self.redis_client:
            await self.redis_client.delete(key)

    async def exists(self, key: str) -> bool:
        """Check if key exists"""
        if self.redis_client:
            return await self.redis_client.exists(key) > 0
        return False 